sqlgen/batch.go

```diff
@@ -0,0 +1,104 @@
+package sqlgen
+
+import (
+	"bytes"
+	"sort"
+)
+
+// makeBatchQuery combines a set of filters into a single SQL that matches a
+// union of the filters.
+//
+// Different filters with equal sets of columns are combined into grouped IN
+// queries. For example, the filters
+//   {"id": 10}
+//   {"id": 20}
+//   {"name": "Bob", "city": "San Francisco"}
+// will get combined to form the query
+//   WHERE ((city, name)) in ("San Francisco", "Bob") OR id in (10, 20)
+// (except with parameter substitution.)
+func makeBatchQuery(filters []Filter) (string, []interface{}) {
+	// A batchQueryGroup holds all values tuples for a given set of columns in the
+	// WHERE statement.
+	type batchQueryGroup struct {
+		columns []string
+		tuples  [][]interface{}
+	}
+
+	// Put every filter in its group.
+	groups := make(map[string]*batchQueryGroup)
+	for _, filter := range filters {
+		// Special case the empty filter: Return an empty WHERE clause.
+		if len(filter) == 0 {
+			return "", nil
```

Comment:
clarified the comment. essentially, an empty filter means matching all rows, and as far as I know you can't say `() IN (())`, so it requires special case logic.